# German Used Cars - Data Cleaning & E.D.A.

In this project we will be taking a closer look at the state of Germany's used car market. Using the dataset that is found on [Kaggle](https://www.kaggle.com/datasets/wspirat/germany-used-cars-dataset-2023), we are going to be cleaning, analysing and exploring the said dataset's data in order to accumulate as much information as possible about this market and come up with answers to various questions.

## Part 0 - Importing the Libraries & Taking a Glance at the Data

In [579]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

raw_df = pd.read_csv('Data - German_Used_Cars.csv')
df = raw_df.copy()

In [580]:
df.head()

Unnamed: 0       brand              model  color registration_date  year  \
0           0  alfa-romeo     Alfa Romeo GTV    red           10/1995  1995   
1           1  alfa-romeo     Alfa Romeo 164  black           02/1995  1995   
2           2  alfa-romeo  Alfa Romeo Spider  black           02/1995  1995   
3           3  alfa-romeo  Alfa Romeo Spider  black           07/1995  1995   
4           4  alfa-romeo     Alfa Romeo 164    red           11/1996  1996   

  price_in_euro power_kw power_ps transmission_type fuel_type  \
0          1300      148      201            Manual    Petrol   
1         24900      191      260            Manual    Petrol   
2          5900      110      150           Unknown    Petrol   
3          4900      110      150            Manual    Petrol   
4         17950      132      179            Manual    Petrol   

  fuel_consumption_l_100km fuel_consumption_g_km  mileage_in_km  \
0            10,9 l/100 km              260 g/km       160500.0   
1                      NaN              - (g/km)       190000.0   
2                      NaN              - (g/km)       129000.0   
3             9,5 l/100 km              225 g/km       189500.0   
4             7,2 l/100 km              - (g/km)        96127.0   

                       offer_description  
0                              2.0 V6 TB  
1                    Q4 Allrad, 3.2L GTA  
2                          ALFA ROME 916  
3                   2.0 16V Twin Spark L  
4  3.0i Super V6, absoluter Topzustand !

In [581]:
df.shape

(251079, 15)

In [582]:
df.columns

Index(['Unnamed: 0', 'brand', 'model', 'color', 'registration_date', 'year',
       'price_in_euro', 'power_kw', 'power_ps', 'transmission_type',
       'fuel_type', 'fuel_consumption_l_100km', 'fuel_consumption_g_km',
       'mileage_in_km', 'offer_description'],
      dtype='object')

In [583]:
df.dtypes

Unnamed: 0                    int64
brand                        object
model                        object
color                        object
registration_date            object
year                         object
price_in_euro                object
power_kw                     object
power_ps                     object
transmission_type            object
fuel_type                    object
fuel_consumption_l_100km     object
fuel_consumption_g_km        object
mileage_in_km               float64
offer_description            object
dtype: object

In [584]:
df.isna().sum()

Unnamed: 0                      0
brand                           0
model                           0
color                         166
registration_date               4
year                            0
price_in_euro                   0
power_kw                      134
power_ps                      129
transmission_type               0
fuel_type                       0
fuel_consumption_l_100km    26873
fuel_consumption_g_km           0
mileage_in_km                 152
offer_description               1
dtype: int64

After our brief glance of the data there are some things that we need to keep in mind moving forward. More specifically:

- There is an extra column, namely 'Unnamed: 0', which does not serve any significant purpose and thus must be *dropped*.
- The majority of the data types in the dataset are objects (in particular **strings**) and should be turned to their appropriate dtypes.
- There is a considerable amount of missing values, compared to the amount of rows in the dataset, and we have to treat them carefully in order not to lose as much data during the 'cleaning' phase.

## Part 1 - Data Cleaning

In order to improve the results of our analysis, we must first go through the stage of cleaning our data; making it more accurate and removing unwanted clutter. In the following steps we will be taking the necessary actions in doing that.

In [585]:
# dropping the columns we won't be using
df.drop(['Unnamed: 0', 'offer_description', 'power_kw'], axis=1, inplace=True)

In [586]:
print('Unique values in the registration_date column before cleaning:')
df.registration_date.unique()

Unique values in the registration_date column before cleaning:


array(['10/1995', '02/1995', '07/1995', '11/1996', '04/1996', '12/1996',
       '07/1996', '01/1996', '02/1996', '10/1996', '05/1997', '06/1997',
       '10/1997', '03/1997', '01/1997', '02/1997', '04/1997', '07/1997',
       '09/1997', '07/1998', '05/1998', '01/1998', '11/1998', '09/1998',
       '07/1999', '02/1999', '06/1999', '12/1999', '05/1999', '01/1999',
       '04/1999', '03/1999', '10/2000', '06/2000', '05/2000', '01/2000',
       '04/2000', '11/2000', '03/2000', '02/2000', '09/2001', '02/2001',
       '03/2001', '01/2001', '11/2001', '12/2001', '04/2001', '05/2001',
       '07/2001', '06/2001', '10/2002', '12/2002', '01/2002', '09/2002',
       '11/2002', '06/2002', '03/2002', '02/2002', '05/2003', '10/2003',
       '07/2003', '02/2003', '01/2003', '03/2003', '08/2003', '06/2003',
       '03/2004', '02/2004', '09/2004', '04/2004', '08/2004', '01/2004',
       '11/2004', '06/2004', '10/2004', '12/2004', '07/2004', '05/2004',
       '10/2005', '06/2005', '07/2005', '09/2005', 

In [587]:
#in order to filter out those unexpected values we create a condition; we want the date to be of the %m/%Y format
# and to only contain numbers

condition = (df.registration_date.str.contains('/') == True) & (df.registration_date.str.match("[0-9]+"))
df[~condition]

brand                      model   color registration_date  \
2400          audi                    Audi A4    grey            Manual   
2743          audi                    Audi A4  silver              8449   
5297          audi                    Audi A1    blue              8999   
9616          audi                    Audi A1   white            Manual   
12206         audi                    Audi A6   black            Manual   
...            ...                        ...     ...               ...   
241462  volkswagen  Volkswagen Passat Variant  silver            Manual   
242569  volkswagen             Volkswagen up!     red         18.200 km   
242579  volkswagen             Volkswagen up!     red          8.500 km   
246113  volkswagen             Volkswagen up!     red            Manual   
247366       volvo                 Volvo XC90    blue         Automatic   

             year              price_in_euro      power_ps transmission_type  \
2400       Petrol            KLIMAAUTOMATIK*            75            Manual   
2743    Automatic                  Automatik      194 g/km         Automatic   
5297       Manual                     6-Gang      126 g/km            Manual   
9616       Petrol   XEN AAC SHZ FSP APS 5TRG            90            Manual   
12206      Diesel     NAVplus LEDER PANO ALC           140            Manual   
...           ...                        ...           ...               ...   
241462     Diesel                   Navi uvm           110            Manual   
242569    06/2021                        PDC  4,9 l/100 km            Manual   
242579    07/2021          Beats-Soundsystem  4,9 l/100 km            Manual   
246113     Petrol                        SHZ            85            Manual   
247366     Petrol                MATTSCHWARZ           235         Automatic   

         fuel_type fuel_consumption_l_100km fuel_consumption_g_km  \
2400       07/2004                 187 g/km                  2004   
2743     74.194 km                   Petrol               11/2005   
5297    110.250 km                   Petrol               04/2011   
9616       06/2014                 128 g/km                  2014   
12206      12/2016                 119 g/km                  2016   
...            ...                      ...                   ...   
241462     06/2020                 104 g/km                  2020   
242569      Manual                     2021                Petrol   
242579      Manual                     2021                Petrol   
246113     03/2023                 112 g/km                  2023   
247366     12/2016                 - (g/km)                  2016   

        mileage_in_km  
2400           2460.0  
2743              NaN  
5297              NaN  
9616          12250.0  
12206         21250.0  
...               ...  
241462        20900.0  
242569            NaN  
242579            NaN  
246113        23590.0  
247366        41999.0  

[199 rows x 12 columns]

As we can see above, most of the values in this subset are misplaced and it would take a lot of effort to bring it back to a proper, usable format. Luckily, since the number of misplaced entries is very low (only 199 out of 251079 total entries) we can safely remove those rows from our dataset.

In [588]:
misplaced_values_list = df[~condition].index.values
df.drop(misplaced_values_list, axis=0, inplace=True)

print('Unique values in the registration_date column after cleaning:')
df.registration_date.unique()

Unique values in the registration_date column after cleaning:


array(['10/1995', '02/1995', '07/1995', '11/1996', '04/1996', '12/1996',
       '07/1996', '01/1996', '02/1996', '10/1996', '05/1997', '06/1997',
       '10/1997', '03/1997', '01/1997', '02/1997', '04/1997', '07/1997',
       '09/1997', '07/1998', '05/1998', '01/1998', '11/1998', '09/1998',
       '07/1999', '02/1999', '06/1999', '12/1999', '05/1999', '01/1999',
       '04/1999', '03/1999', '10/2000', '06/2000', '05/2000', '01/2000',
       '04/2000', '11/2000', '03/2000', '02/2000', '09/2001', '02/2001',
       '03/2001', '01/2001', '11/2001', '12/2001', '04/2001', '05/2001',
       '07/2001', '06/2001', '10/2002', '12/2002', '01/2002', '09/2002',
       '11/2002', '06/2002', '03/2002', '02/2002', '05/2003', '10/2003',
       '07/2003', '02/2003', '01/2003', '03/2003', '08/2003', '06/2003',
       '03/2004', '02/2004', '09/2004', '04/2004', '08/2004', '01/2004',
       '11/2004', '06/2004', '10/2004', '12/2004', '07/2004', '05/2004',
       '10/2005', '06/2005', '07/2005', '09/2005', 

Now, let's take the time to address the missing values in our dataframe. During our brief look at the dataset at the beginning of the project we noted that most of them are located in the Fuel Consumption column. Before we proceed with any data manipulation, let's take a moment to question why some of these values are missing.

- Electric vehicles do not consume fuel, but instead rely on a battery to move.
- The dataset is a collection of online car listings, thus many entries are subject to human lack of knowledge. The fuel consumption entry is no exception as many car owners may not be aware of such details about their car.

In [589]:
missing_values_df = pd.DataFrame(df[df.fuel_consumption_l_100km.isna()]['fuel_type'].value_counts())
missing_values_df.columns = ['Missing Values']
missing_values_df['Total Rows'] = df['fuel_type'].value_counts()
missing_values_df['% of Missing Values'] = round(missing_values_df['Missing Values'] \
/missing_values_df['Total Rows']*100, 2)
missing_values_df.sort_values('% of Missing Values', ascending=False)

Missing Values  Total Rows  % of Missing Values
Electric                 5496        5967                92.11
Unknown                    73          96                76.04
Hydrogen                   44          82                53.66
Other                      91         178                51.12
Ethanol                     5          10                50.00
Diesel Hybrid              99         476                20.80
CNG                        94         508                18.50
Hybrid                   2257       12607                17.90
LPG                       185        1255                14.74
Diesel                   7425       86421                 8.59
Petrol                  11104      143280                 7.75

What meaning does the above table convey to us exactly? We can see that roughly 92% of the total rows of electric vehicles have a missing value in the Fuel Consumption column. Dropping these rows from our dataset now would be disastrous for our later on analysis. On the other hand, we can drop the rows for which the Fuel Type is 'Unknown' or 'Other' because not only are they significantly less in size, compared to the total row count, but also due to the fact that the ratio between the missing and total entries is still high.

So, for now, let's simply replace the missing values of all the fuel types with zero and drop the entries which we do not need.

In [590]:
df = df[df['fuel_type'] != 'Unknown']
df = df[df['fuel_type'] != 'Other']

In [591]:
# replacing the nan values with zero for the rest of the entries and printing the results
df['fuel_consumption_l_100km'] = df.loc[:,'fuel_consumption_l_100km'].fillna(0)
na_fuel_consumption_rows = df['fuel_consumption_l_100km'].isna().sum()
print('Number of rows with missing values in the Fuel Consumption column after cleaning: ', na_fuel_consumption_rows)


Number of rows with missing values in the Fuel Consumption column after cleaning:  0


We still need to address the zero-valued rows for the electric vehicles which we replaced in the previous step, so let's do that now.

In [592]:
df[df.fuel_type == 'Electric']

brand        model  color registration_date  year price_in_euro  \
16552    audi  Audi e-tron  beige           09/2019  2019         51888   
16559    audi  Audi e-tron  beige           07/2019  2019         53990   
16561    audi  Audi e-tron  beige           11/2019  2019         54870   
16571    audi  Audi e-tron  beige           12/2019  2019         61989   
16579    audi  Audi e-tron   blue           02/2019  2019         32930   
...       ...          ...    ...               ...   ...           ...   
251033  volvo    Volvo C40  black           05/2023  2023         52890   
251037  volvo   Volvo XC40  black           04/2023  2023         49900   
251048  volvo    Volvo C40  black           01/2023  2023         51990   
251056  volvo    Volvo C40  black           05/2023  2023         60520   
251078  volvo   Volvo XC40   gold           03/2023  2023         52900   

       power_ps transmission_type fuel_type fuel_consumption_l_100km  \
16552       408         Automatic  Electric                        0   
16559       408         Automatic  Electric                        0   
16561       408         Automatic  Electric                        0   
16571       408         Automatic  Electric                        0   
16579       408         Automatic  Electric                        0   
...         ...               ...       ...                      ...   
251033      231         Automatic  Electric                        0   
251037      231         Automatic  Electric                        0   
251048      231         Automatic  Electric                        0   
251056      231         Automatic  Electric                        0   
251078      218         Automatic  Electric                        0   

       fuel_consumption_g_km  mileage_in_km  
16552      359 km Reichweite        84800.0  
16559      359 km Reichweite        51000.0  
16561                 0 g/km        82814.0  
16571                 0 g/km        55990.0  
16579      359 km Reichweite        84300.0  
...                      ...            ...  
251033     400 km Reichweite            8.0  
251037                0 g/km        14900.0  
251048                0 g/km         2106.0  
251056     400 km Reichweite         3000.0  
251078     438 km Reichweite           50.0  

[5967 rows x 12 columns]

The column named 'fuel_consumption_g_km' contains information about the cars' battery range (*Reichweite = Range* in German) so we can extract that information and make a new column with the battery range values.

In [593]:
df['range'] = '' #creating an empty column
el_vehicles = df[df.fuel_type == 'Electric'].index.values #index values of all electric vehicles

df.loc[el_vehicles, 'range'] = df.loc[el_vehicles, 'fuel_consumption_g_km'].str.extract(pat='(\d+)', expand=False)

In [599]:
#dropping unnecessary columns and rearranging
df = df[['brand', 'model', 'color', 'registration_date', 'year', 'power_ps', \
'transmission_type', 'fuel_type', 'fuel_consumption_l_100km', 'range', 'mileage_in_km', 'price_in_euro']]

df

brand              model  color registration_date  year power_ps  \
0       alfa-romeo     Alfa Romeo GTV    red           10/1995  1995      201   
1       alfa-romeo     Alfa Romeo 164  black           02/1995  1995      260   
2       alfa-romeo  Alfa Romeo Spider  black           02/1995  1995      150   
3       alfa-romeo  Alfa Romeo Spider  black           07/1995  1995      150   
4       alfa-romeo     Alfa Romeo 164    red           11/1996  1996      179   
...            ...                ...    ...               ...   ...      ...   
251074       volvo         Volvo XC40  white           04/2023  2023      261   
251075       volvo         Volvo XC90  white           03/2023  2023      235   
251076       volvo          Volvo V60  white           05/2023  2023      197   
251077       volvo         Volvo XC40  white           05/2023  2023      179   
251078       volvo         Volvo XC40   gold           03/2023  2023      218   

       transmission_type fuel_type fuel_consumption_l_100km range  \
0                 Manual    Petrol            10,9 l/100 km         
1                 Manual    Petrol                        0         
2                Unknown    Petrol                        0         
3                 Manual    Petrol             9,5 l/100 km         
4                 Manual    Petrol             7,2 l/100 km         
...                  ...       ...                      ...   ...   
251074         Automatic    Hybrid                        0         
251075         Automatic    Diesel             7,6 l/100 km         
251076         Automatic    Diesel             4,7 l/100 km         
251077         Automatic    Hybrid                        0         
251078         Automatic  Electric                        0   438   

        mileage_in_km price_in_euro  
0            160500.0          1300  
1            190000.0         24900  
2            129000.0          5900  
3            189500.0          4900  
4             96127.0         17950  
...               ...           ...  
251074         1229.0         57990  
251075         4900.0         89690  
251076         1531.0         61521  
251077         1500.0         57890  
251078           50.0         52900  

[250606 rows x 12 columns]

Since most of our columns still contain string based values, it would be beneficial to convert them into their appropriate data type.